Dockerfile

```diff
@@ -1,16 +1,86 @@
-#
-# To build this Docker image: docker build -t fpm .
-#
-# To run this Docker container interactively: docker run -it fpm
-#
-FROM alpine:3.7
-
-RUN apk add --no-cache \
-        ruby \
-        ruby-dev \
-        gcc \
-        libffi-dev \
-        make \
-        libc-dev \
-        rpm \
-    && gem install --no-ri --no-rdoc fpm
+# Are we running against the minimal container, or the everything
+# container? Minimal is mostly the compiled package tools. Everything
+# pulls in scripting langauges.
+ARG BASE_ENV=everything
+
+# Are we running tests, or a release? Tests build and run against the
+# CWD, where release will use the downloaded gem.
+ARG TARGET=test
+
+# Container to throw an error if called with a bare `docker build .`
+FROM ubuntu:18.04 as error
+RUN echo "\n\n\nHey! Use buildkit. See the Makefile or docs\n\n\n"
+RUN false
+
+# Base container is used for various release and test things
+FROM ubuntu:18.04 as minimal-base
+
+# Runtime deps. Build deps go in the build or test containers
+RUN apt-get update \
+	&& apt-get -y dist-upgrade \
+	&& apt-get install --no-install-recommends -y \
+	ruby rubygems rubygems-integration \
+	bsdtar \
+	cpio \
+	debsigs \
+	pacman \
+	rpm  \
+	squashfs-tools \
+	xz-utils \
+	zip \
+	&& rm -rf /var/lib/apt/lists/* \
+	&& apt-get clean
+
+# everything container includes all the scripting languages. These
+# greatly embiggen the underlying docker container, so they're
+# conditionalized.
+FROM minimal-base AS everything-base
+RUN apt-get update \
+	&& apt-get install --no-install-recommends -y \
+	cpanminus \
+	npm \
+	perl \
+	python3-pip \
+	&& pip3 --no-cache-dir install setuptools \
+	&& pip3 --no-cache-dir install wheel \
+	&& pip3 --no-cache-dir install virtualenv virtualenv-tools3 \
+	&& update-alternatives --install /usr/bin/python python /usr/bin/python3 10 \
+	&& rm -rf /var/lib/apt/lists/*
+
+
+# Run tests against the current working directory. This is a bit
+# orthogonal to the container release process, but it has a lot of
+# same dependancies, so we reuse it. This uses COPY to allow rspect to
+# initall the gems, but runtime usage expects you to mount a volume
+# into /src
+FROM ${BASE_ENV}-base AS test
+WORKDIR /src
+RUN apt-get update \
+	&& apt-get install --no-install-recommends -y \
+	gcc make ruby-dev libc-dev lintian git
+# installing ffi here is a bit of an optimization for how COPY and layer reuse works
+RUN gem install --no-ri --no-rdoc bundler ffi
+RUN mkdir /origsrc
+COPY . /origsrc
+RUN cd /origsrc && bundle install
```

Comment:
This step fails for me. I'm not sure what's causing it since bundler is rather _obviously_ installed in the previous step.

```
 => CACHED [test 2/6] RUN apt-get update  && apt-get install --no-install  0.0s
 => CACHED [test 3/6] RUN gem install --no-ri --no-rdoc bundler ffi        0.0s
 => CACHED [test 4/6] RUN mkdir /origsrc                                   0.0s
 => [test 5/6] COPY . /origsrc                                             0.1s
 => ERROR [test 6/6] RUN cd /origsrc && bundle install                     0.5s
------
 > [test 6/6] RUN cd /origsrc && bundle install:
#13 0.471 /usr/lib/ruby/2.5.0/rubygems.rb:289:in `find_spec_for_exe': can't find gem bundler (>= 0.a) with executable bundle (Gem::GemNotFoundException)
#13 0.471 	from /usr/lib/ruby/2.5.0/rubygems.rb:308:in `activate_bin_path'
#13 0.471 	from /usr/local/bin/bundle:23:in `<main>'
------
executor failed running [/bin/sh -c cd /origsrc && bundle install]: exit code: 1
make: *** [Makefile:61: .docker-test-everything] Error 1
```

Reply:
If I make it list the bundler version before trying to run it, it does show that bundler is installed, and yet, `bundle install` itself fails inside of bundler? somehow? I've never seen this behavior before.
```
#14 [test 7/7] RUN cd /origsrc; gem list bundler;  bundle install
#14 sha256:152d7cbdd81380115de7ebb36522f118b1b0095aa4b642ab2c803eb3e3a03760
#14 0.689 bundler (2.2.25)
#14 0.748 /usr/lib/ruby/2.5.0/rubygems.rb:289:in `find_spec_for_exe': can't find gem bundler (>= 0.a) with executable bundle (Gem::GemNotFoundException)
```

Reply:
Oh, I got it sorted. It's _extremely_ confusing due to the error message. This blog post explains it: https://bundler.io/blog/2019/05/14/solutions-for-cant-find-gem-bundler-with-executable-bundle.html

I'll work on a fix.